### ISSUE TITLE
Keyboard drop target navigation shouldn't be dependent on render order

### ISSUE THREAD

**Issue by LFDanLu** [adobe/react-spectrum#2071]
TODO: ask accessibility for feedback on what the actual behavior should be

# 🐛 Bug Report 

Update: Some work has been done that addresses the original issue and some other points listed here. Repurposing for a high level issue:

At the moment, DragManager `validDropTargets` registers targets via a useLayoutEffect, meaning the order is dependent on when things render. Ideally, we would look at the element's position in the DOM to get a logical order in which hitting Tab would move you from drop target to drop target. This would also make it so that drop elements that are added later wouldn't be punted to the last in the order

Old info:
When the user starts a drag operation via hitting Enter on a draggable element, the first drop target that gets focused is not always the closest droppable element. If there are nested droppable elements nested within each other, the deepest child is the initial focused drop target. This is because child elements useLayoutEffect occurs before parent's useLayoutEffect, resulting in the child element registering itself as a drop target first via https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/dnd/src/useDrop.ts#L195. The DragManager then takes the [first droppable target](https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/dnd/src/DragManager.ts#L351-L355) to focus when [dragging starts](https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/dnd/src/DragManager.ts#L70-L76).

## 🤔 Expected Behavior

The drop target order should be sorted to follow DOM order

## 😯 Current Behavior

The drop target order is set by render order

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | 3.27
| Browser          |
| Operating System |

## 🧢 Your Company/Team

RSP

**Comment by LFDanLu** [comment]
Some initial progress/potential fix on https://github.com/adobe/react-spectrum/tree/issue_2071. 
Bit iffy on the filter change, also investigating the ordering if a new droppable element is added after initial render

**Comment by devongovett** [comment]
I think I fixed this in my work on the docs.

**Comment by LFDanLu** [comment]
Retested, verified that the closest droppable element is focused first when a drag is started via keyboard but the tab order is somewhat unexpected: https://reactspectrum.blob.core.windows.net/reactspectrum/cd0376c480ef7f554ea0aebac6669f28fa87740a/storybook/index.html?path=/story/drag-and-drop--nested-drop-regions&providerSwitcher-toastPosition=bottom. If you trigger keyboard dnd in the story about and hit tab after focus moves to the outer drop target, focus returns to the draggable item rather than going to the nested drop target.

**Comment by LFDanLu** [comment]
Repurposing ticket for drop target navigation order refactor
